jax/_src/typing/simple.py

```diff
@@ -0,0 +1,34 @@
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Simple type definitions, not """
+
+# This file should avoid importing from jax in order to avoid circular imports.
+from __future__ import annotations
+
+from typing import Any, Sequence, Union
+from typing_extensions import Protocol
+import numpy as np
+
+class HasDtypeAttribute(Protocol):
+  dtype: Dtype
```

Comment:
Must this protocol supports writeable `dtype` variable?
`numpy.typing` counterpart has read only `dtype` variable(property) https://github.com/numpy/numpy/blob/6094eff95c19c433648593f3cb6bca489396f9ae/numpy/_typing/_dtype_like.py#L83-L87. 

[PEP 544: By default, protocol variables as defined above are considered readable and writable. To define a read-only protocol variable, one can use an (abstract) property](https://peps.python.org/pep-0544/#protocol-members)

Reply:
Also, name it as `SupportsDType`?

Reply:
Hmmm, I guess we can just use `numpy.typing.DTypeLike` here since we use `np.dtype(x)` for every `DTypeLike` `x`.

Reply:
`numpy.typing.DTypeLike` is too permissive (e.g. it allows structured types) see https://jax.readthedocs.io/en/latest/jep/12049-type-annotations.html#inputs-should-be-permissively-typed

Reply:
I got the point! But it seems that JAX use `np.dtype(xxx)` everywhere?

Reply:
Yes it does. Invalid inputs would generate runtime errors later in the process.

Reply:
Thanks for the suggestions - changes in #12813